pytorch_tabular/models/base_model.py

```diff
@@ -159,38 +169,69 @@ def calculate_metrics(self, y, y_hat, tag):
                 prog_bar=True,
             )
         return metrics
-    
+
     def data_aware_initialization(self, datamodule):
         pass
 
-    @abstractmethod
+    def compute_backbone(self, x: Dict):
+        # Returns output
+        x = self.backbone(x)
+        return x
+
+    def compute_head(self, x: Tensor):
+        y_hat = self.head(x)
+        if (self.hparams.task == "regression") and (
+            self.hparams.target_range is not None
+        ):
+            for i in range(self.hparams.output_dim):
+                y_min, y_max = self.hparams.target_range[i]
+                y_hat[:, i] = y_min + nn.Sigmoid()(y_hat[:, i]) * (y_max - y_min)
+        if type(self.head) == nn.Identity:
+            return {"logits": y_hat}
+        else:
+            return {"logits": y_hat, "backbone_features": x}
+
+    def compute_ssl_head(self, x: Dict):
+        return getattr(ssl, self.hparams.ssl_task)(input_dim=self.backbone.output_dim)(x)
+
     def forward(self, x: Dict):
-        pass
+        x = self.compute_backbone(x)
+        if self.hparams.task == "ssl":
+            return self.compute_ssl_head(x)
+        return self.compute_head(x)
 
     def predict(self, x: Dict, ret_model_output: bool = False):
+        assert self.hparams.task != "ssl", "It's not allowed to use the method predict in case of ssl task"
         ret_value = self.forward(x)
         if ret_model_output:
             return ret_value.get("logits"), ret_value
         else:
             return ret_value.get("logits")
 
+    def shared_step(self, batch):
+        if self.hparams.task == "ssl":
+            y = self(batch)["logits"]
+            batch_augmented = getattr(augmentations, self.hparams.aug_task)(batch)
+            y_hat = self(batch_augmented)["logits"]
+        else:
+            y = batch["target"]
+            y_hat = self(batch)["logits"]
+        return y, y_hat
+
     def training_step(self, batch, batch_idx):
-        y = batch["target"]
-        y_hat = self(batch)["logits"]
+        y, y_hat = self.shared_step(batch)
```

Comment:
Can we change the function name to something more intuitive and indicative of the action being done in the method? something like forward_pass or something like that?

Reply:
Fixed